interfaces/connection.go

```diff
@@ -0,0 +1,222 @@
+		return copyRecursive(val)
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+func (plug *ConnectedPlug) StaticAttrs() (map[string]interface{}, error) {
+	return copyAttributes(plug.plugInfo.Attrs)
+}
+
+func (plug *ConnectedPlug) Attr(key string) (interface{}, error) {
+	if plug.dynamicAttrs != nil {
+		if val, ok := plug.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return plug.StaticAttr(key)
+}
+
+func (plug *ConnectedPlug) Attrs() (map[string]interface{}, error) {
+	if plug.dynamicAttrs != nil {
+		return plug.dynamicAttrs, nil
+	}
+	return nil, fmt.Errorf("dynamic attributes not initialized")
+}
+
+func (plug *ConnectedPlug) SetAttr(key string, value interface{}) error {
+	if plug.dynamicAttrs == nil {
+		plug.dynamicAttrs = make(map[string]interface{})
+	}
+	if _, ok := plug.plugInfo.Attrs[key]; ok {
+		return fmt.Errorf("attribute %q cannot be overwritten", key)
+	}
+	plug.dynamicAttrs[key] = value
+	return nil
+}
+
+func (slot *ConnectedSlot) Interface() string {
+	return slot.slotInfo.Interface
+}
+
+func (slot *ConnectedSlot) Name() string {
+	return slot.slotInfo.Name
+}
+
+func (slot *ConnectedSlot) Snap() *snap.Info {
+	return slot.slotInfo.Snap
+}
+
+func (slot *ConnectedSlot) Apps() map[string]*snap.AppInfo {
+	return slot.slotInfo.Apps
+}
+
+func (slot *ConnectedSlot) SecurityTags() []string {
+	return slot.slotInfo.SecurityTags()
+}
+
+func (slot *ConnectedSlot) StaticAttr(key string) (interface{}, error) {
+	if val, ok := slot.slotInfo.Attrs[key]; ok {
+		return copyRecursive(val)
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+func (slot *ConnectedSlot) StaticAttrs() (map[string]interface{}, error) {
+	return copyAttributes(slot.slotInfo.Attrs)
+}
+
+func (slot *ConnectedSlot) Attr(key string) (interface{}, error) {
+	if slot.dynamicAttrs != nil {
+		if val, ok := slot.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return slot.StaticAttr(key)
+}
+
+func (slot *ConnectedSlot) Attrs() (map[string]interface{}, error) {
+	if slot.dynamicAttrs != nil {
+		return slot.dynamicAttrs, nil
+	}
+	return nil, fmt.Errorf("dynamic attributes not initialized")
+}
+
+func (slot *ConnectedSlot) SetAttr(key string, value interface{}) error {
+	if slot.dynamicAttrs == nil {
+		slot.dynamicAttrs = make(map[string]interface{})
+	}
+	if _, ok := slot.slotInfo.Attrs[key]; ok {
+		return fmt.Errorf("attribute %q cannot be overwritten", key)
+	}
+	slot.dynamicAttrs[key] = value
+	return nil
+}
+
+func (conn *Connection) Interface() string {
+	return conn.plug.plugInfo.Interface
+}
+
+func copyAttributes(value map[string]interface{}) (map[string]interface{}, error) {
+	cpy, err := copyRecursive(value)
+	if err != nil {
+		return nil, err
+	}
+	return cpy.(map[string]interface{}), err
+}
+
+func copyRecursive(value interface{}) (interface{}, error) {
+	switch v := value.(type) {
+	case string:
+		return v, nil
+	case bool:
+		return v, nil
+	case int:
+		return int64(v), nil
+	case int64:
```

Comment:
The typed customization here feels a bit strange. This must have been cleaned up on the way in, by whatever changed these values. We shouldn't have to reparse it here. The goal of this is only to copy it over to prevent mutation, and as such we don't need to know the individual types unless the type is mutable, which is only the case for the two entries below. All the above can be in default. 

We also need a comment in the code that does prevent other types from going in that this code must be changed if new mutable types are introduced. The lack of synchronicity will create very unfortunate runtime issues, that we'll only know about from users when they decide to use it in a particular interface.

Reply:
yes, this seems to come from the code we have in snap/info_snap_yaml.go to indeed normalize values, but at this point they should be normalized